torch/_inductor/comms.py

```diff
@@ -342,6 +342,148 @@ def reorder_compute_and_comm_for_overlap(
+
+    # Find all graph inputs and their resize counts
+    graph_input_to_resized_to_full_node_idxes = defaultdict(list)
+    graph_input_to_resized_to_0_node_idxes = defaultdict(list)
+    for idx, node in enumerate(node_list):
+        if (
+            node.op == "call_function"
+            and node.target == torch.ops.inductor.resize_storage_bytes_.default
+            and node.args[0].op == "placeholder"
+        ):
+            graph_input = node.args[0]
+            new_size = node.args[1]
+            if new_size > 0:
+                graph_input_to_resized_to_full_node_idxes[graph_input].append(idx)
+            else:
+                graph_input_to_resized_to_0_node_idxes[graph_input].append(idx)
+
+    # Find all eligible unsharded params and their corresponding graph intermediates.
+    unsharded_param_to_fsdp_copy_node_idxes = defaultdict(list)
+    for idx, node in enumerate(node_list):
+        if node.op == "call_function" and node.target == torch.ops.fsdp.copy_.default:
+            fsdp_copy_node = node
+            unsharded_param = node.args[0]
+            assert (
+                unsharded_param.op == "placeholder"
+            ), "Assumed all FSDP2 `unsharded_param`s to be graph input, but it's not true!"
+            # An unsharded param is eligible for this pass only if it's resized to 0 at the end of graph.
+            num_resize_to_full = len(
+                graph_input_to_resized_to_full_node_idxes[unsharded_param]
+            )
+            num_resize_to_0 = len(
+                graph_input_to_resized_to_0_node_idxes[unsharded_param]
+            )
+            last_resize_to_full_node_idx = (
+                graph_input_to_resized_to_full_node_idxes[unsharded_param][-1]
+                if num_resize_to_full > 0
+                else -1
+            )
+            last_resize_to_0_node_idx = (
+                graph_input_to_resized_to_0_node_idxes[unsharded_param][-1]
+                if num_resize_to_0 > 0
+                else -1
+            )
+            if (
+                num_resize_to_full <= num_resize_to_0
+            ) and last_resize_to_0_node_idx > last_resize_to_full_node_idx:
+                unsharded_param_to_fsdp_copy_node_idxes[unsharded_param].append(idx)
+
+    # Check no user mutation on any unsharded_param
+    def is_allowed_mutation(node):
+        return (
+            node.target == torch.ops.fsdp.copy_.default
+            or node.target == torch.ops.inductor.resize_storage_bytes_.default
+        )
+
+    for node in node_list:
+        if (
+            node.op == "call_function"
+            and isinstance(node.target, torch._ops.OpOverload)
+            and node.target._schema.is_mutable
+            and not is_allowed_mutation(node)
+            and any(arg in unsharded_param_to_fsdp_copy_node_idxes for arg in node.args)
+        ):
+            raise RuntimeError(
+                "User mutation on FSDP2 unsharded param is not allowed when Traceable FSDP2 is used"
```

Comment:
Discussed offline: eager FSDP2 doesn't really support mutations on params.

But also as discussed below:
> But if we/someone adds a new mutable op into the graph in the future, we probably want this pass to raise an error instead of starting to be silently wrong.

So we will still need to check for mutable ops (added below).